clouddriver-dcos/src/main/groovy/com/netflix/spinnaker/clouddriver/dcos/security/DcosCredentialsInitializer.groovy

```diff
@@ -54,33 +39,54 @@ class DcosCredentialsInitializer implements CredentialsInitializerSynchronizable
                                                           CatsModule catsModule,
                                                           ApplicationContext applicationContext,
                                                           AccountCredentialsRepository accountCredentialsRepository,
                                                           List<ProviderSynchronizerTypeWrapper> providerSynchronizerTypeWrappers) {
 
     // TODO what to do with clouddriverUserAgentApplicationName?
+    Map<String, DcosConfigurationProperties.Cluster> clusterMap = new HashMap<>()
+
+    for (DcosConfigurationProperties.Cluster cluster in dcosConfigurationProperties.clusters) {
+      clusterMap.put(cluster.name, cluster)
+    }
 
     def (ArrayList<DcosConfigurationProperties.Account> accountsToAdd, List<String> namesOfDeletedAccounts) =
     ProviderUtils.calculateAccountDeltas(accountCredentialsRepository,
-      DcosCredentials,
+      DcosAccountCredentials,
       dcosConfigurationProperties.accounts)
 
     accountsToAdd.each { DcosConfigurationProperties.Account account ->
       try {
+        List<DcosClusterCredentials> clusterCredentials = new ArrayList<>()
+
+        for (DcosConfigurationProperties.ClusterCredentials clusterConfig in account.clusters) {
+          DcosConfigurationProperties.Cluster cluster = clusterMap.get(clusterConfig.name)
+
+          if (cluster == null) {
+            LOGGER.warn("Cluster [${cluster.name}] could not be found in the list of configured clusters for account [${account}]")
+            continue
+          }
+
+          def key = DcosClientCompositeKey.buildFromVerbose(account.name, cluster.name)
+
+          if (!key.isPresent()) {
+            LOGGER.warn("Something went wrong with the composite key creation for account [${account.name}] and cluster [${cluster.name}]")
+            continue
+          }
+
+          def dockerRegistries = cluster.dockerRegistries ? cluster.dockerRegistries : account.dockerRegistries
 
-        def dcosCredentials = DcosCredentials.builder().name(account.name)
-          .environment(account.environment)
-          .accountType(account.accountType)
-          .dcosUrl(account.dcosUrl)
-          .dockerRegistries(account.dockerRegistries)
-          .requiredGroupMembership(account.requiredGroupMembership)
-          .secretStore(account.secretStore)
-          .dcosClientConfig(DcosCredentialsInitializer.buildConfig(account))
-          .build()
+          clusterCredentials.add(DcosClusterCredentials.builder().key(key.get()).dcosUrl(cluster.dcosUrl)
+                  .secretStore(cluster.secretStore).dockerRegistries(dockerRegistries)
+                  .dcosConfig(DcosConfigurationProperties.buildConfig(account, cluster, clusterConfig)).build())
+        }
 
+        def dcosCredentials = new DcosAccountCredentials(account.name, account.environment, account.accountType, account.dockerRegistries, account.requiredGroupMembership, clusterCredentials)
```

Comment:
Construct using the builder instead of the constructor (which should be made private), otherwise we're skipping validation.